### PR TITLE
check db file permission

### PR DESCRIPTION
**Pull request by millken**
fix crash when the db file cant be access

### CODE REVIEW

File: blockchain/blockdao/blockdao.go [hunk 1]
```diff
@@ -75,6 +75,7 @@ type (
 func NewBlockDAO(indexers []BlockIndexer, cfg config.DB) BlockDAO {
 	blkStore, err := filedao.NewFileDAO(cfg)
 	if err != nil {
+		log.L().Fatal(err.Error(), zap.Any("cfg", cfg))
```
Comment (Madongming) [inline blockchain/blockdao/blockdao.go:78]: Can you judge the type of error?

File: blockchain/filedao/filedao.go [hunk 2]
```diff
@@ -37,6 +37,7 @@ var (
 // vars
 var (
 	ErrFileNotExist     = errors.New("file does not exist")
+	ErrFileAccess       = errors.New("file does not access")
```
Comment (dustinxie) [inline blockchain/filedao/filedao.go:40]: ErrFileCantAccess = errors.New("cannot access file")

File: blockchain/filedao/filedao_util.go [hunk 3]
```diff
@@ -45,6 +46,11 @@ func readFileHeader(filename, fileType string) (*FileHeader, error) {
 		return nil, ErrFileNotExist
 	}
 
+	err := syscall.Access(filename, syscall.O_RDWR)
+	if err != nil {
+		return nil, ErrFileAccess
+	}
+
```
Comment (dustinxie) [inline blockchain/filedao/filedao_util.go:53]: combine line 43 to 52 into:

```go
fileExists() error
```

File: blockchain/filedao/filedao_util.go [hunk 4]
```diff
@@ -80,12 +74,16 @@ func readFileHeader(filename, fileType string) (*FileHeader, error) {
 	}
 }
 
-func fileExists(name string) (int64, bool) {
+func fileExists(name string) error {
 	info, err := os.Stat(name)
 	if err != nil || info.IsDir() {
-		return 0, false
+		return ErrFileNotExist
```
Comment (dustinxie) [inline blockchain/filedao/filedao_util.go:80]: `if err != nil || info.IsDir() || info.Size == 0`

File: blockchain/filedao/filedao_util.go [hunk 5]
```diff
@@ -40,15 +40,9 @@ func checkMasterChainDBFile(defaultName string) (*FileHeader, error) {
...
 
 func readFileHeader(filename, fileType string) (*FileHeader, error) {
-	size, exist := fileExists(filename)
-	if !exist || size == 0 {
-		// default chain db file does not exist
-		return nil, ErrFileNotExist
-	}
-
-	err := syscall.Access(filename, syscall.O_RDWR)
+	err := fileExists(filename)
```
Comment (dustinxie) [inline blockchain/filedao/filedao_util.go:43]: combine into if below

**Review by dustinxie — DISMISSED** [review]

File: chainservice/chainservice.go [hunk 6]
```diff
@@ -385,6 +385,9 @@ func New(
 			return nil, err
 		}
 	}
+	if dao == nil {
+		return nil, errors.New("blockdao is nil")
+	}
```
Comment (dustinxie) [inline chainservice/chainservice.go:390]: why need this?

for test, we may have NewBlockDAO() return nil
for actual case like "make run", it should return a non-nil blockdao?

**Review by dustinxie — APPROVED** [review]

**Pull request merged**